bigquery_etl/cli/query.py

```diff
@@ -534,3 +536,183 @@ def initialize(name, sql_dir, project_id, dry_run):
+                    click.echo(
+                        "A new column has been added to "
+                        f"{project_name}.{dataset_name}.{table_name}"
+                    )
+                else:
+                    click.echo(
+                        f"The schema for {project_name}.{dataset_name}.{table_name} "
+                        "is unchanged"
+                    )
+
+
+@schema.command(help="Deploy the query schema", name="validate")
+@click.argument("name")
+@sql_dir_option
+@click.option(
+    "--project-id",
+    "--project_id",
+    help="GCP project ID",
+    default="moz-fx-data-shared-prod",
+    callback=is_valid_project,
+)
+def validate_schema(name, sql_dir, project_id):
+    """Validate the defined query schema against the query and destination table."""
+    if not is_authenticated():
+        click.echo(
+            "Authentication to GCP required. Run `gcloud auth login` "
+            "and check that the project is set correctly."
+        )
+        sys.exit(1)
+    client = bigquery.Client()
+
+    query_files = _queries_matching_name_pattern(name, sql_dir, project_id)
+
+    for query_file in query_files:
+        query_file_path = Path(query_file)
+
+        if not (query_file_path.parent / "schema.json").is_file():
+            click.echo(f"No schema defined for {query_file_path}", err=True)
+        else:
+            table_name = query_file_path.parent.name
+            dataset_name = query_file_path.parent.parent.name
+            project_name = query_file_path.parent.parent.parent.name
+
+            click.echo(f"Validate schema for {query_file_path}")
+            with open(query_file_path) as f:
+                sql = f.read()
+            config = bigquery.QueryJobConfig(
+                dry_run=True,
+                default_dataset=f"{project_name}.{dataset_name}",
+                query_parameters=[
+                    bigquery.ScalarQueryParameter(
+                        "submission_date", "DATE", "2000-01-01"
+                    ),
+                ],
+            )
+            query_schema = client.query(sql, config)._job_statistics()["schema"]
+
+            with open(query_file_path.parent / "schema.json") as schema_file:
+                defined_schema = json.dumps(json.load(schema_file), sort_keys=True)
+
+                if query_schema != defined_schema:
+                    click.echo(
+                        f"Schema defined in {query_file_path.parent / 'schema.json'} "
+                        f"incompatible with query {query_file_path}",
+                        err=True,
+                    )
+
+                table = client.get_table(f"{project_name}.{dataset_name}.{table_name}")
+                table_schema = io.StringIO("")
+                client.schema_to_json(table.schema, table_schema)
+
+                if defined_schema != json.dumps(
+                    json.loads(table_schema.getvalue()), sort_keys=True
+                ):
```

Comment:
There is probably a better way to compare schemas here

Reply:
A complete solution would probably be to traverse the schema files, descending into structs where necessary, building a list of field names in dotted notation (`client_info.client_id`, etc.). We would be able to compare alphabetically sorted lists to determine if the new schema is missing any existing fields, and we'd be able to report what new fields are added. This technique would also conveniently ignore the complication of having column comments not exist at the query level.

Hmm... this would also need to take into account data type to make sure there's not an incompatible change.

Reply:
:amiyaguchi has definitely done some work around this for the safety-checks for MSG, using an approach I think similar to what :klukas describes.

One thing to note here is that the bigquery API sometimes returns legacy SQL for schemas and it can make diffing a lot more complicated: https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/bigquery_table#schema. This is perhaps not an issue for query schemas where legacy sql not used.